### PR TITLE
feat: commit, repo, and merge request mock updates

### PR DESCRIPTION
**Pull request by jasonwoods-7**
* use project id when target project id is not specified on a new merge request
* implement get raw blob on repository client
* create action handler should create directory if it does not exist
* throw `GitLabNotFoundException` when directory does not exist to replicate what gitlab does
* add id and mode when retrieving tree items
* use default user when commit author & email not specified

### CODE REVIEW

File: NGitLab.Mock/Repository.cs [hunk 1]
```diff
@@ -204,12 +204,12 @@ public Commit Commit(CommitCreate commitCreate)
         var mustCreateBranch = !repo.Commits.Any();
         if (!mustCreateBranch)
         {
-            Commands.Checkout(repo, commitCreate.Branch);
+            Commands.Checkout(repo, commitCreate.StartBranch ?? commitCreate.StartSha);
```
Comment (louis-z) [inline NGitLab.Mock/Repository.cs]: There are problems with this change. But to be fair, there were already issues with the existing mock code...

I just added a few more commit creation tests in the main branch, to document GitLab's behavior. We should try to reproduce that behavior in the mocks as much as possible. @jasonwoods-7, could you rebase your branch, look up the aforementioned tests and adapt your PR accordingly? Similar tests should be added for the mocks.

**Review by louis-z — CHANGES_REQUESTED** [review]

File: NGitLab.Mock/Repository.cs (hunk 1 above)
Comment (jasonwoods-7) [reply NGitLab.Mock/Repository.cs]: ok, i think i have all the cases covered now. let me know if i missed any.

File: NGitLab.Mock/Repository.cs [hunk 2]
```diff
@@ -558,14 +584,27 @@ internal string FetchBranchFromFork(Project fork, string branch)
         return $"remotes/{forkRemote.Name}/{branch}";
     }
 
-    private static Tree GetTreeItem(string repositoryPath, string filePath)
+    private Tree GetTreeItem(string repositoryPath, string filePath)
```
Comment (jasonwoods-7) [inline NGitLab.Mock/Repository.cs]: I'm working on refactoring this method to more accurately match what GitLab does.

Comment (jasonwoods-7) [reply NGitLab.Mock/Repository.cs]: done

File: NGitLab.Mock.Tests/UsingVerify/RepositoryMockTests.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
+﻿using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using NGitLab.Mock.Clients;
+using NGitLab.Mock.Config;
+using NUnit.Framework;
+using RepositoryGetTreeOptions = NGitLab.Models.RepositoryGetTreeOptions;
+using NThrows = NUnit.Framework.Throws;
```
Comment (louis-z) [inline NGitLab.Mock.Tests/UsingVerify/RepositoryMockTests.cs]: You should be able to use the following alias instead:

```cs
using Throws = NUnit.Framework.Throws;
```

File: NGitLab.Mock/Repository.cs [hunk 4]
```diff
@@ -201,15 +201,35 @@ public Commit Commit(User user, string message, string targetBranch, IEnumerable
...
-        var mustCreateBranch = !repo.Commits.Any();
-        if (!mustCreateBranch)
+        var branchExists = repo.Branches.Any(b => string.Equals(b.FriendlyName, commitCreate.Branch, StringComparison.Ordinal));
+
+        if (branchExists)
+        {
+            throw new GitLabBadRequestException($"A branch called '{commitCreate.Branch}' already exists.");
+        }
+
+        var isBranchEmpty = !repo.Commits.Any();
```
Comment (louis-z) [inline NGitLab.Mock/Repository.cs]: It seems to me `isRepoEmpty` would be a more appropriate name here.

**Review by louis-z — CHANGES_REQUESTED** [review]

File: NGitLab.Mock.Tests/UsingVerify/RepositoryMockTests.cs (hunk 3 above)
Comment (jasonwoods-7) [reply NGitLab.Mock.Tests/UsingVerify/RepositoryMockTests.cs]: updated

File: NGitLab.Mock/Repository.cs (hunk 4 above)
Comment (jasonwoods-7) [reply NGitLab.Mock/Repository.cs]: updated

**Pull request merged**